### PR TITLE
fix(acceleration): stop a refresh already running from answering a later waiter (refs #13544)

### PR DESCRIPTION
**Pull request by claudespice**
## Summary

A refresh-completion waiter was satisfied by the **next completion recorded on the table, whichever refresh produced it**. Nothing correlated the refresh a caller triggered with the completion it was woken by, so a refresh that was *already running* when the caller took its waiter could finish a moment later and release it — reporting on work that started before the caller changed anything.

On the partition-assignment path that is a wrong answer rather than a late one. `Runtime::update_partition_refresh_sql` installs new partition filters, triggers a refresh to load them, and broadcasts `PartitionsLoaded` when the wait returns:

1. A periodic refresh **P** is running.
2. A partition assignment change lands. The executor installs the new filters, takes a completion waiter, and triggers refresh **T** to load the newly assigned partitions.
3. **P** finishes and records a completion. The waiter resolves.
4. `PartitionsLoaded` is broadcast — before **T**, the refresh carrying the new filters, has run.

The scheduler is now told the executor holds partitions whose rows it has not read, and routes queries for them there. The result is silently incomplete query results, not an error.

The window is not hypothetical: the refresher's external-trigger arm sleeps for refresh jitter **before** it hands the request on, and that sleep blocks the same `select!` loop that drains completions — so a completion of **P** landing during the jitter is queued and processed immediately after **T** is dispatched, which is exactly the interleaving above.

### Root cause

`RefreshCompletion` counted completions. A `next()` waiter was decided by the `watch` version it subscribed at, which distinguishes *before the waiter* from *after the waiter* but not *which refresh*. Those are different questions whenever a refresh is in flight.

## Changes

- **`RefreshCompletion` numbers each request.** `issue()` hands out a `RefreshRequestId` and `record(id)` reports that request's refresh complete. A `next()` waiter remembers the highest id issued when it was taken and resolves only above it; `any()` — the initial-load question — keeps its lenient semantics as a threshold of 0.
- **Numbering happens at *issue*, not at start.** A caller installs its change, takes the waiter, then triggers, so every id issued afterwards belongs to a refresh that will read the change. A refresh issued between the change and the waiter is merely not credited — conservative, never the other way round.
- **The id travels with the request through `RefreshTaskRunner` and comes back with the completion.** This is required rather than bookkeeping: a request arriving mid-refresh causes the `select!` to drop the running future, so that refresh is cancelled and never reports. A completion therefore cannot be attributed by counting requests against completions — the counts do not line up.
- **The CDC apply loop records without an id** (`record_untriggered()`). It answers no trigger, so it keeps precisely the semantics it had: every waiter taken before it is released, and none taken after.
- Ids saturate rather than wrap, because a waiter now compares against them. The ceiling — 2^64 refreshes of one table in one process — is unreachable, and reaching it would strand a waiter rather than release one early, so the failure direction stays "a readiness ack that is never sent" rather than "one sent for data that was never loaded".

### Behavioural note

A `next()` waiter can now wait one refresh longer than before in exactly the case this fixes: where an in-flight refresh used to release it, it now waits for the refresh the caller actually triggered. That is the point — the earlier release was the bug. Liveness is preserved because `trigger_refresh` uses `send().await` on the trigger channel, so a successful trigger always produces exactly one issued id, and a superseded request is always followed by one with a higher id.

`#13603` — revalidating a completion against a table that may have been removed or rebuilt while it ran — is a different property of the same signal and stays open; the doc comment on `RefreshCompletionOutcome` is narrowed to reference only that remaining half.

### Why `refs #13544` and not `fixes`

This closes the interleaving #13544 works through — a refresh **already running** when the waiter was taken can no longer answer it — but it does not establish the general property that issue's opening paragraph states, that a waiter is correlated with *the* trigger its caller sent. Two windows survive, both raised by Copilot on this PR:

- A request **enqueued before** the waiter but not yet numbered — the refresher numbers it after its jitter sleep — gets an id above the waiter's threshold, so its completion answers.
- A request enqueued **after** the waiter answers by construction, which is correct for an unrestricted refresh (it runs after the caller's change and reloads it) but wrong for one carrying restrictive `RefreshOverrides`, which may not load what the caller was waiting for.

Numbering at enqueue rather than at dispatch would close the first. It does not close the second, and the second is the one that needs a decision: binding a waiter to exactly one request means a request that is **superseded** — the runner drops the running future when a new request arrives, so it never reports — must resolve its waiters as something other than "answered", and each acting caller has to decide what supersession means for it. That is a design question rather than a bug fix, and worth a maintainer's view, so #13544 stays open owning it rather than being auto-closed by a merge that only narrows it.

## Test plan

- `make lint`
- `make nextest`

New tests in `crates/runtime-table/src/accelerated/refresh_completion.rs`:

- `next_ignores_the_completion_of_a_refresh_requested_before_it` — the regression. A request issued before the waiter completes while the caller's own request is still running; the waiter must not resolve. Fails on the old code, where the completion released it.
- `next_is_answered_by_the_refresh_requested_after_it` — the other half of the same scenario.
- `next_is_answered_by_a_later_request_when_the_earlier_one_never_completes` — liveness for the cancelled in-flight refresh.
- `record_never_walks_the_threshold_backwards` — an out-of-order completion must not retract an answered wait.
- `any_is_satisfied_by_the_completion_of_a_refresh_requested_before_it` — correlation narrows `next` only.
- `any_waits_while_a_request_is_only_issued`, `a_request_left_in_flight_by_a_drop_is_abandoned` — an issued request is not a completion.
- `an_untriggered_completion_releases_the_waiters_taken_before_it`, `an_untriggered_completion_does_not_release_a_later_next_waiter` — the CDC arm.
- `a_waiter_resolves_at_the_top_of_the_id_range` — replaces the wrap test, which measured a property the comparison removed.

End-to-end through the real refresher, in `refresh.rs`:

- `test_a_refresh_already_running_does_not_answer_a_later_waiter` — the interleaving, driven rather than timed. A refresh over a `MemTable` completes inside one scheduler pass, so a waiter taken "while it runs" is really taken after it; a first attempt that polled a request counter tripped its own precondition assertion instead of passing vacuously, which is what sent this to a gate. `GatedSource` holds the source scan open and reports when one is entered, so the waiters are provably taken with the refresh in flight, and the test asserts that precondition (`completed_requests() == 0`) rather than assuming it. It then releases the gate, waits for that refresh to complete, asserts it did *not* answer the waiter, triggers a second refresh and asserts that one does. No sleeps and no spin loops.

The pre-existing `#13086` tests (`test_completion_taken_before_a_trigger_survives_the_window`, `test_initial_load_flag_is_stored_before_the_completion_is_recorded`) are the guard on the wiring: if the runner reported the wrong id, they hang rather than pass.

`scripts/check_table_layers.py` gains one `ALLOWED` entry. `GatedSource` wraps a `TableProvider` and implements one, which the guard refuses because a layer walk stops at such a type silently; it is a test double that holds a scan open rather than part of a dataset's layer stack, so it joins the six test doubles already listed there.

One unrelated test-helper change rides along, and does **not** fix anything on its own: `wait_until_ready_status` read the `dataset_load_state` gauge series positionally (`get_metric()[0]`), and that gauge is labelled by dataset, so the series it reads depends on which datasets other tests in the binary registered. This PR adds one (`gated`), which makes the positional read strictly more fragile than before, so it now selects by the `dataset` label. `test_refresh_status_change_to_ready` **already fails on unmodified `trunk`** under `cargo test -p runtime-table --lib accelerated::refresh` — measured at `4c70e06533` with the working tree reset to `origin/trunk` — for a different reason: the gauge is a process-wide `LazyLock` that binds to whichever meter provider a sibling test installed first. That is filed as #13708 and is not addressed here. The module is green under `cargo nextest run -p runtime-table` (356/356), which is what `make nextest` runs, because nextest gives each test its own process.

## Review gate

- Adversarial review: **skipped** — `codex` ran and exited 1 with `Codex error: Your workspace is out of credits. Ask your workspace owner to refill in order to continue.`; `grok` is not installed in this environment (receipt: `engine=none exit=1 job=none`). No engine was available, so this was not run. In its place the approach was re-examined directly against the failure modes it exists to catch — cancellation liveness, trigger-channel back-pressure, the `any()`/CDC arms, and the saturation ceiling — and each is covered by a test listed above.
- `/security-review`: no findings. The diff adds an in-process `u64` counter and a tuple on two internal `tokio::mpsc` channels; no new parsing, deserialization, path, SQL, or credential surface, and the one new log field is an integer request id on a line that already rendered the refresh result. The change strictly narrows when a readiness ack fires, so both degenerate cases leave a waiter unresolved rather than acking early.
- `/simplify`: run on the resulting diff. One fix applied — `record_untriggered` looks like avoidable duplication of `issue`, so the reason it is a single `send_modify` rather than `record(self.issue())` is now stated in its doc: the intermediate state of that pair is an issued-but-uncompleted request, and a waiter taken there would skip the completion and wait for the following apply. Reviewed and deliberately left: no existing sequence helper to reuse; `issue`'s captured-`mut` is the correct way to read a value out of `send_modify` atomically; moving `issued` to a separate atomic would stop `issue` waking waiters but splits one invariant across two primitives to optimise a cold path. Behaviour unchanged, no test weakened, no snapshot touched.

### Neuters

Each guard was reverted in turn and the suite re-run, to show the tests fail without the fix rather than alongside it:

| Neuter | Result |
| --- | --- |
| The refresher records an uncorrelated completion (`record_untriggered()` in place of `record(request_id)`) — i.e. the pre-fix behaviour | `test_a_refresh_already_running_does_not_answer_a_later_waiter` fails, and only that test |
| `next()` loses its threshold (always 0) | `next_ignores_the_completion_of_a_refresh_requested_before_it`, `next_ignores_a_completion_recorded_before_the_waiter_was_taken` and `an_untriggered_completion_does_not_release_a_later_next_waiter` all fail |
| `record` assigns instead of `max` | `record_never_walks_the_threshold_backwards` fails |
| The CDC arm issues without recording | the pre-existing `test_start_changes_stream_signals_dataset_ready` fails |

The runner's id echo is load-bearing by construction: the integration test sends 7 and 11, so a runner echoing a constant, or the wrong one of two in flight, fails at least one assertion.

Refs #13544

Refs #13708

## Checklist

- [x] Regression test that fails on the old code and passes on the new
- [x] Edge cases: cancelled in-flight request, out-of-order completion, drop mid-request, untriggered (CDC) completion, id ceiling
- [x] No `.unwrap()` outside tests; no new lint suppressions
- [x] `cargo fmt --all`

### CODE REVIEW

**Comment by claudespice** [comment]
@copilot review

File: crates/runtime-table/src/accelerated/refresh_completion.rs [hunk 1]
```diff
@@ -111,22 +180,37 @@ impl RefreshCompletion {
...
-            receiver.mark_changed();
+        // Read the state back through the receiver rather than the sender: a
+        // transition landing between subscribing and reading is either seen here
+        // or wakes the receiver, never dropped between them.
+        let receiver = self.state.subscribe();
+        let threshold = if accept_earlier {
+            0
+        } else {
+            receiver.borrow().issued
+        };
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime-table/src/accelerated/refresh_completion.rs:210]: Correct, and this changed the PR rather than the code.

You are right that the threshold does not bind a waiter to the request its caller enqueued. Ids are allocated after the refresher dequeues a trigger and finishes its jitter sleep, so a request accepted *before* the waiter can be numbered *after* it and answer. When that request carries restrictive `RefreshOverrides`, its completion means something narrower than the waiter asked for, and `PartitionsLoaded` can still go out for rows nobody loaded.

Numbering at enqueue and carrying the id through the external-trigger channel closes that first window, and I nearly did it here. It does not close the second one: a request enqueued *after* the waiter also answers by construction, which is right for an unrestricted refresh and wrong for an overridden one. Binding to exactly one ticket is what closes both, and it needs the piece you flagged in parentheses — a request that is **superseded** never completes at all, because the runner's `select!` drops the running future when a new request arrives, so an exactly-targeted waiter would hang rather than resolve. Making that safe means the runner has to report supersession and each acting caller has to decide what it means for it (`update_partition_refresh_sql` broadcasting readiness and `load_view` creating a schedule do not obviously want the same answer).

That is a design decision rather than a bug fix, so rather than land a partial correlation under a closing keyword, I have retargeted this PR from `fixes #13544` to `refs #13544`. What it does close is the interleaving #13544 works through — a refresh **already running** when the waiter was taken can no longer answer it, which is the case the partition path actually hits, since the jitter sleep blocks the same `select!` that drains completions. #13544 stays open owning the remainder, with both windows you named written into the PR body so the next person starts from them rather than rediscovering them.

File: crates/runtime-table/src/accelerated/refresh.rs [hunk 2]
```diff
@@ -1085,12 +1086,17 @@ impl Refresher {
                             sleep(Self::compute_delay(Duration::from_secs(0), Some(max_jitter))).await;
                         }
 
-                        if let Err(err) = start_refresh.send(overrides_opt).await {
+                        // Numbered here rather than at the trigger: a caller
+                        // takes its waiter before triggering, so an id issued
+                        // now is one the waiter has not seen and a refresh
+                        // already running cannot claim.
+                        let request_id = issue_refresh_request(refresh_completion.as_ref());
+                        if let Err(err) = start_refresh.send((request_id, overrides_opt)).await {
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime-table/src/accelerated/refresh.rs:1094]: Correct, and this changed the PR rather than the code.

You are right that the threshold does not bind a waiter to the request its caller enqueued. Ids are allocated after the refresher dequeues a trigger and finishes its jitter sleep, so a request accepted *before* the waiter can be numbered *after* it and answer. When that request carries restrictive `RefreshOverrides`, its completion means something narrower than the waiter asked for, and `PartitionsLoaded` can still go out for rows nobody loaded.

Numbering at enqueue and carrying the id through the external-trigger channel closes that first window, and I nearly did it here. It does not close the second one: a request enqueued *after* the waiter also answers by construction, which is right for an unrestricted refresh and wrong for an overridden one. Binding to exactly one ticket is what closes both, and it needs the piece you flagged in parentheses — a request that is **superseded** never completes at all, because the runner's `select!` drops the running future when a new request arrives, so an exactly-targeted waiter would hang rather than resolve. Making that safe means the runner has to report supersession and each acting caller has to decide what it means for it (`update_partition_refresh_sql` broadcasting readiness and `load_view` creating a schedule do not obviously want the same answer).

That is a design decision rather than a bug fix, so rather than land a partial correlation under a closing keyword, I have retargeted this PR from `fixes #13544` to `refs #13544`. What it does close is the interleaving #13544 works through — a refresh **already running** when the waiter was taken can no longer answer it, which is the case the partition path actually hits, since the jitter sleep blocks the same `select!` that drains completions. #13544 stays open owning the remainder, with both windows you named written into the PR body so the next person starts from them rather than rediscovering them.

**Review by lukekim — APPROVED** [review]
